app/models/collection.rb

```diff
@@ -459,20 +460,34 @@ def reveal_authors!
   end
 
   def reveal_collection_items
-    approved_collection_items.each { |collection_item| collection_item.update_attribute(:unrevealed, false) }
+    collection_items.each { |collection_item| collection_item.update_attribute(:unrevealed, false) }
     send_reveal_notifications
   end
 
   def reveal_collection_item_authors
-    approved_collection_items.each { |collection_item| collection_item.update_attribute(:anonymous, false) }
-  end
-
+    collection_items.each { |collection_item| collection_item.update_attribute(:anonymous, false) }
+  end
+
+  # A work is not marked unrevealed until the user has approved its inclusion
+  # in a collection, so we don't want to send notifications unless 
+  # user_approval_status is approved.
+  #
+  # TODO: When AO3-2240 is fixed, we should ideally only send notifications for
+  # items that had their unrevealed status changed from true to false when the 
+  # collection's unrevealed setting was changed. With the current code, it's
+  # possible to send notifications twice for the same work if the moderator
+  # reveals the collection, sets it back to unrevealed (which does not and
+  # should not set existing collection items back to unreveald), and then
```

Comment:
"unrevealed"

Reply:
I'll get this one in the next PR.

Reply:
And when I say I'll get it, I mean the fix for AO3-2240 is to remove this method entirely and delete a couple of lines from `notify_of_status_change` in the collection item model.

Reply:
Just noting for myself that's not PRed because it'll need to be updated once this is merged. It's here, though: https://github.com/otwcode/otwarchive/compare/master...sarken:AO3-2240_final?expand=1

Reply:
I'm not sure where to put this comment, because the PR hasn't been submitted yet, but since this and AO3-2240 seem to be treated as two halves of a whole, I figured I'd mention it here:

In the code for AO3-2240 that you've linked to, the recipient notification email will probably be very susceptible to stale data issues. Because `UserMailer#recipient_notification` is checking whether the work is unrevealed, and the recipient notification email is queued up right after the work is changed from unrevealed to revealed, any kind of delay in reads will cause the "A Gift Work For You" email to go missing.

Reply:
Hmm, that's what I was trying to avoid by having it check the collection item rather than the work. How else might I get around it?

Reply:
The collection item and the work are updated in the same transaction, so checking one vs. the other won't make a difference in production. (It does make a difference in the test environment, because delayed jobs are performed synchronously,`notify_of_status_change` is an `after_update` hook, and `update_work` is an `after_save` hook. But it won't help with stale data issues.)

I'm not really sure what the best approach is.

You could make `notify_of_status_change` an `after_update_commit` hook, which would probably handle most of the intermittent email failures. But I'm not sure it'd fix them all. Additionally, `after_commit` hooks with `saved_change_to_attribute?` checks make me nervous, because if an object is saved multiple times in the same transaction, you may not get consistent results.

You could also try to move the check out of `UserMailer#recipient_notification` and into one of the functions that calls it (as long as it's within an `after_update` or `after_save` hook). Though that could run into problems if the commit is rolled back after the notification is enqueued, because then recipients would be notified about gifts that they can't see. Maybe you could use the `after_commit_everywhere` gem for that?

Alternatively, you could try to add a significant delay to the emails, like subscription emails. The subscription code can afford to check whether the works are anonymous because subscription emails are only sent once an hour.

(Also, I noticed another potential issue: It looks like it might send notifications even if the `CollectionItem` isn't `approved_by_user`.)

Reply:
Opened a PR here for discussion: https://github.com/otwcode/otwarchive/pull/3672

Would the delay need to be an hour? I think a five or ten minute delay on all gift notifications would be fine, even beneficial -- it would give the creator time to correct or remove the recipient info -- but an hour seems less than ideal.

>Also, I noticed another potential issue: It looks like it might send notifications even if the CollectionItem isn't approved_by_user.

The check in the mailer should handle that, I think? (The code is most of a year old so I'm a little fuzzy on it.) `approved_collection_items` should be checking both collection and user approval status.

Reply:
A five or ten minute delay would probably be fine.

> The check in the mailer should handle that, I think? (The code is most of a year old so I'm a little fuzzy on it.) `approved_collection_items` should be checking both collection and user approval status.

The check in the mailer ensures that you'll only get notifications if the work is visible. But if someone has a work that's already visible, with an unrevealed CollectionItem that's not `approved_by_user`, the recipient will get a notification when that CollectionItem is revealed, even though they (probably) already received a notification when the work first became visible.